src/operator/numpy/np_unique_op.h

```diff
@@ -0,0 +1,185 @@
+    }
+  }
+};
+
+struct UniqueComputeMaskGPUKernel {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int64_t i,
+                                  int64_t* out_data,
+                                  const DType* in_data,
+                                  const int64_t numel) {
+    if (i == 0) {
+      out_data[i] = 1;
+    } else {
+      out_data[i] = 0;
+      for (int64_t j = 0; j < numel; ++j) {
+        if (in_data[(i - 1) * numel + j] != in_data[i * numel + j]) {
+          out_data[i] = 1;
+          break;
+        }
+      }
+    }
+  }
+};
+
+struct UniqueReturnInverseKernel {
+  MSHADOW_XINLINE static void Map(int i,
+                                  int64_t* unique_inverse,
+                                  const int32_t* prefix_sum,
+                                  const int64_t* perm) {
+      dim_t j = perm[i];
+      unique_inverse[j] = prefix_sum[i] - 1;
+  }
+};
+
+struct UniqueReturnCountsKernel {
+  MSHADOW_XINLINE static void Map(int i, int64_t* unique_counts, const int32_t* idx) {
+      unique_counts[i] = idx[i + 1] - idx[i];
+  }
+};
+
+struct NumpyUniqueParam : public dmlc::Parameter<NumpyUniqueParam> {
+  bool return_index, return_inverse, return_counts;
+  dmlc::optional<int> axis;
+  DMLC_DECLARE_PARAMETER(NumpyUniqueParam) {
+    DMLC_DECLARE_FIELD(return_index)
+    .set_default(false)
+    .describe("If true, return the indices of the input.");
+    DMLC_DECLARE_FIELD(return_inverse)
+    .set_default(false)
+    .describe("If true, return the indices of the input.");
+    DMLC_DECLARE_FIELD(return_counts)
+    .set_default(false)
+    .describe("If true, return the number of times each unique item appears in input.");
+    DMLC_DECLARE_FIELD(axis)
+    .set_default(dmlc::optional<int>())
+    .describe("An integer that represents the axis to operator on.");
+  }
+};
+
+inline bool NumpyUniqueType(const nnvm::NodeAttrs& attrs,
+                            std::vector<int> *in_attrs,
+                            std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 1U);
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, in_attrs->at(0));
+  TYPE_ASSIGN_CHECK(*in_attrs, 0, out_attrs->at(0));
+  for (size_t i = 1; i < out_attrs->size(); ++i) {
+    TYPE_ASSIGN_CHECK(*out_attrs, i, mshadow::kInt64);
+  }
+  return out_attrs->at(0) != -1;
+}
+
+inline bool NumpyUniqueStorageType(const nnvm::NodeAttrs& attrs,
+                            const int dev_mask,
+                            DispatchMode* dispatch_mode,
+                            std::vector<int> *in_attrs,
+                            std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 1U);
+  // CHECK_EQ(out_attrs->size(), 1U);
+  for (int &attr : *in_attrs) {
+    CHECK_EQ(attr, kDefaultStorage) << "Only default storage is supported";
+  }
+  for (int &attr : *out_attrs) {
+    attr = kDefaultStorage;
+  }
+  *dispatch_mode = DispatchMode::kFComputeEx;
+  return true;
+}
+
+template<typename xpu>
```

Comment:
Since cpu and gpu do not share anything in implementation, a template declaration is not necessary. You use different function names for cpu and gpu FCompute registration.